pkg/runner/action.go

```diff
@@ -381,63 +382,123 @@ func execAsComposite(step actionStep) common.Executor {
+			env[k] = eval.Interpolate(v)
 		}
-		// Workaround end
-		compositerc.Composite = action
-		envToEvaluate := mergeMaps(compositerc.Env, step.getStepModel().Environment())
-		compositerc.Env = make(map[string]string)
-		// origEnvMap: is used to pass env changes back to parent runcontext
-		origEnvMap := make(map[string]string)
-		for k, v := range envToEvaluate {
-			ev := eval.Interpolate(v)
-			origEnvMap[k] = ev
-			compositerc.Env[k] = ev
+
+		// run with the global config but without secrets
+		configCopy := *rc.Config
+		configCopy.Secrets = nil
+
+		// create a run context for the composite action to run in
+		compositerc := &RunContext{
+			Name:    rc.Name,
+			JobName: rc.JobName,
+			Run: &model.Run{
+				JobID: "composite-job",
+				Workflow: &model.Workflow{
+					Name: rc.Run.Workflow.Name,
+					Jobs: map[string]*model.Job{
+						"composite-job": {},
+					},
+				},
+			},
+			Config:           &configCopy,
+			StepResults:      map[string]*model.StepResult{},
+			JobContainer:     rc.JobContainer,
+			Inputs:           inputs,
+			ActionPath:       containerActionDir,
+			ActionRepository: rc.ActionRepository,
+			ActionRef:        rc.ActionRef,
+			Env:              env,
+			Masks:            rc.Masks,
 		}
-		compositerc.Inputs = inputs
-		compositerc.ExprEval = compositerc.NewExpressionEvaluator()
 
-		err := compositerc.compositeExecutor()(ctx)
+		ctx = WithCompositeLogger(ctx, &compositerc.Masks)
+
+		// We need to inject a composite RunContext related command
+		// handler into the current running job container
+		// We need this, to support scoping commands to the composite action
+		// executing.
+		rawLogger := common.Logger(ctx).WithField("raw_output", true)
+		logWriter := common.NewLineWriter(compositerc.commandHandler(ctx), func(s string) bool {
+			if rc.Config.LogOutput {
+				rawLogger.Infof("%s", s)
+			} else {
+				rawLogger.Debugf("%s", s)
+			}
+			return true
+		})
+		oldout, olderr := compositerc.JobContainer.ReplaceLogWriter(logWriter, logWriter)
+		defer (func() {
+			rc.JobContainer.ReplaceLogWriter(oldout, olderr)
+		})()
+
+		err := compositerc.compositeExecutor(action)(ctx)
 
-		// Map outputs to parent rc
-		eval = compositerc.NewStepExpressionEvaluator(step)
+		// Map outputs from composite RunContext to job RunContext
+		eval = compositerc.NewExpressionEvaluator()
 		for outputName, output := range action.Outputs {
-			backup.setOutput(ctx, map[string]string{
+			rc.setOutput(ctx, map[string]string{
 				"name": outputName,
 			}, eval.Interpolate(output.Value))
 		}
 
-		backup.Masks = append(backup.Masks, compositerc.Masks...)
-		// Test if evaluated parent env was altered by this composite step
-		// Known Issues:
-		// - you try to set an env variable to the same value as a scoped step env, will be discared
-		for k, v := range compositerc.Env {
-			if ov, ok := origEnvMap[k]; !ok || ov != v {
-				backup.Env[k] = v
-			}
-		}
+		rc.Masks = compositerc.Masks
```

Comment:
I forget the extrapath here too, while creating my initial PR. No reviewer complained :(.
```suggestion
		rc.Masks = compositerc.Masks
		rc.ExtraPath = compositerc.ExtraPath
```